workflow/controller/dag.go

```diff
@@ -232,6 +232,21 @@ func (woc *wfOperationCtx) executeDAG(nodeName string, tmplCtx *templateresoluti
 	// kick off execution of each target task asynchronously
 	for _, taskName := range targetTasks {
 		woc.executeDAGTask(dagCtx, taskName)
+
+		// It is possible that target tasks are not reconsidered (i.e. executeDAGTask is not called on them) once they are
```

Comment:
Do you think we can move out the `runOnExitNode` from `executeDAGTask` and handle it here, So OnExitNode logic will be one place instead of checking multiple places?

Reply:
No. This only loops through `targetTasks`, not all tasks. We recurse into the non-leaf nodes through `executeDAGTask` where the other `onExitNode` logic is.

Reply:
I think `runOnExitNode` may not run if `task` is already visited. 

Do you think if you arrange the code like this will fix the problem? In this case, there is no duplicate code and `runOnExitNode` logic is in one place.  

```
func (woc *wfOperationCtx) executeDAGTask(dagCtx *dagContext, taskName string) {
	node := dagCtx.getTaskNode(taskName)
	task := dagCtx.GetTask(taskName)
	if node != nil && node.Fulfilled() {
		if node.Completed() {
			// Run the node's onExit node, if any.
			hasOnExitNode, onExitNode, err := woc.runOnExitNode(task.OnExit, task.Name, node.Name, dagCtx.boundaryID, dagCtx.tmplCtx)
			if hasOnExitNode && (onExitNode == nil || !onExitNode.Fulfilled() || err != nil) {
				// The onExit node is either not complete or has errored out, return.
				return
			}
		}
		return
	}
	
	if _, ok := dagCtx.visited[taskName]; ok {
		return
	}
	dagCtx.visited[taskName] = true
```

Reply:
> I think runOnExitNode may not run if task is already visited.

This is not the case. Every task will be visited once during each operation and will have a chance to run an `onExit` node if it is Fulfilled, Completed, and has an `onExit` node declared.

> Do you think if you arrange the code like this will fix the problem?

No. The issue is that a target task my call `executeTemplate` below and finish the DAG from that call, ensuring that there are no more calls to `executeDAGTask` (and therefore `runOnExitNode`). Rearranging the code in the way you suggested will not solve this and would actually be worse because there would be more unnecessary calls to `runOnExit` node.

Reply:
Because we do "short-circuiting" we will need the `runOnExit` code in at least two places. The candidates are (1) where I have added it now, or (2) following `executeTempalte` near the end of this function. I decided on (1), because (2) would make unnecessary calls to `runOnExitNode` without any benefit.